spock-specs/src/test/groovy/spock/util/concurrent/PollingConditionsSpec.groovy

```diff
@@ -148,18 +148,18 @@ class PollingConditionsSpec extends Specification {
   }
 
   def "correctly handles checks that take longer than given check interval"() {
-    when:
+    when: "polling condition has a small timeout"
     def condition = new PollingConditions(timeout: 0.05)
     boolean secondAttempt = false
-    then:
+
+    then: "if first attempt already takes longer than the timeout, there will be no second one"
```

Comment:
That intention of the test was to check the added block:
```
final long timeout = Math.min(currDelay, start + timeoutMillis - System.currentTimeMillis());
        if (timeout > 0) {
```
that guards against a negative timeout value. 

With the proposed changed algorithm I don't think that we can reliably reproduce a negative value making the test obsolete.

Anyway, these changes to the test are not really testing what you intended, since `eventually` only retries when an exception occurs. As you assert `!secondAttempt` which passes, there won't be any retry making the `sleep 100` unnecessary.

Reply:
No, the test is named `"correctly handles checks that take longer than given check interval"`, and this is what it does now after, irrespective of inner code structure of class under test. It checks behaviour instead of trying to cover a condition shielding the code from a negative value for some field or variable:

During the first attempt the closure sets `secondAttempt = true`. If a second attempt was started (which should not occur because the maximum time has already elapsed after the first attempt), this time `assert !secondAttempt` would fail. The fact that this does not occur and the test passes, verifies the expected behaviour.

Reply:
Right, sorry. I need to assert `secondAttempt` instead of `!secondAttempt` in order to make the condition fail during the first try and potentially trigger a second one (which then never occurs because the maximum time has elapsed).